### PR TITLE
Introduce Nginx and Github Actions

### PR DESCRIPTION
**Pull request by pricem14pc**
### What is the context of this PR?
- introduce an Nginx web server and Docker build to return the `service.html` page for any requests served by the container
- introduce a Github Action PR workflow to build the image and push to Dockerhub and GAR

### How to review 
- follow the README ensuring it makes sense and that the steps detailed build a functioning Nginx container
- ensure the container returns the appropriate responses for requests
- ensure the PR image is built and pushed to Dockerhub and GAR

### CODE REVIEW

File: nginx/conf.d/default.conf [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        return 200 '{"status": "Unavailable"}';
+    }
+
+    location / {
+        root /usr/share/nginx/html;
+        index service.html;
+    }
+
+    error_page 404 500 502 503 504 /service.html;
+}
```
Comment (parimooa) [inline nginx/conf.d/default.conf]: Terminating New line missing

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,3 @@
+FROM nginx
+COPY static-html /usr/share/nginx/html
+COPY nginx/conf.d/default.conf /etc/nginx/conf.d/
```
Comment (parimooa) [inline Dockerfile]: Terminating New line missing

File: .github/workflows/pull_request.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          echo "Pushing to DockerHub with tag $TAG"
+          docker push onsdigital/eq-runner-maintenance-page:$TAG
+      - name: Push to GAR
+        env:
+          GAR_SERVICE_KEY: ${{ secrets.GAR_SERVICE_KEY }}
+        run: |
+          echo $GAR_SERVICE_KEY | docker login -u _json_key --password-stdin https://${{ secrets.GAR_LOCATION }}
+          gcloud auth configure-docker ${{ secrets.GAR_LOCATION }}
+          echo "Pushing to GAR with tag $TAG"
+          docker push ${{ secrets.GAR_LOCATION }}/${{ secrets.GAR_PROJECT_ID }}/docker-images/eq-runner-maintenance-page:$TAG
```
Comment (parimooa) [inline .github/workflows/pull_request.yml]: Terminating New line missing

File: .github/workflows/pull_request.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Set Tag and SHA
+        run: |
+          CLEAN_TAG=$(echo "${{ github.event.pull_request.head.ref  }}"  | tr / -)
+          echo "TAG=$CLEAN_TAG" >> $GITHUB_ENV
+          echo "SHA=${{ github.event.pull_request.head.sha }}" >> $GITHUB_ENV
+      - name: Build
+        run: >
+          docker build -t onsdigital/eq-runner-maintenance-page:$TAG
+          -t ${{ secrets.GAR_LOCATION }}/${{ secrets.GAR_PROJECT_ID }}/docker-images/eq-runner-maintenance-page:$TAG .
+      - name: Push to Docker Hub
```
Comment (parimooa) [inline .github/workflows/pull_request.yml]: why do we need to push docker hub ?

File: nginx/conf.d/default.conf [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    server_name _;
+    server_tokens off;
+
+    if ($http_x_forwarded_proto = "http") {
+        return 301 https://$host$request_uri;
+    }
+
+    location /status {
+        default_type application/json;
+        return 200 '{"status": "Unavailable"}';
```
Comment (parimooa) [inline nginx/conf.d/default.conf:12]: Does it need to send `json` response?

**Review by parimooa — COMMENTED** [review]
LGTM, docker image and container working ok. Few minor comments

File: .github/workflows/pull_request.yml (hunk 4 above)
Comment (pricem14pc) [reply .github/workflows/pull_request.yml]: Historically we make all our publicly available ONSdigital product images available via the dockerhub repo. In reality we no longer deploy them into any environments from Dockerhub. It debatable whether we need to continue to push here for the maintenance page, it was done for consistency with the Benchmark and Runner.

File: nginx/conf.d/default.conf (hunk 5 above)
Comment (pricem14pc) [reply nginx/conf.d/default.conf:12]: Runner returns the application version on this route. The Runner uptime checks use this endpoint. There is an argument that when the holding page is up we should be generating uptime failure (which I'd considered). However, my thoughts are; the deployment of the maintenance page will be a controlled and deliberate action so we don't need notification that Runner is down. But more importantly, we need upcheck on the maintenance page itself and continuing to use the `/status` endpoint and the existing notification makes sense.

[EDIT] Following testing in GCP with the associated deploy PR, the upcheck is still triggering as its simply looking for the `OK` in the response. This alert is triggering as this response now contains `Unavailable`.

File: nginx/conf.d/default.conf (hunk 1 above)
Comment (pricem14pc) [reply nginx/conf.d/default.conf]: Done

File: Dockerfile (hunk 2 above)
Comment (pricem14pc) [reply Dockerfile]: Done

File: .github/workflows/pull_request.yml (hunk 3 above)
Comment (pricem14pc) [reply .github/workflows/pull_request.yml]: Done

**Review by parimooa — APPROVED** [review]

File: .github/workflows/pull_request.yml (hunk 4 above)
Comment (MebinAbraham) [inline .github/workflows/pull_request.yml]: I think if we want to keep pushing to Docker hub, then this should also happen on merges to Master to be consistent.

File: nginx/conf.d/default.conf [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    }
+
+    location / {
+        root /usr/share/nginx/html;
+        index service.html;
+    }
+
+    error_page 404 500 502 503 504 /service.html;
```
Comment (MebinAbraham) [inline nginx/conf.d/default.conf:20]: Not sure if we need anything but 404, but not a blocker since it doesn't hurt.

**Review by MebinAbraham — COMMENTED** [review]
Looks good, minor comment.

File: .github/workflows/pull_request.yml (hunk 4 above)
Comment (pricem14pc) [reply .github/workflows/pull_request.yml]: On balance, and taking @parimooa comment into consideration, I'm not sure there is much point in pushing to Dockerhub at all. Unlike Runner we won't have a docker-compose for this image (for a quickstart for example) and therefore have less need for it to be available in Dockerhub. I will remove this step.

File: nginx/conf.d/default.conf (hunk 6 above)
Comment (pricem14pc) [reply nginx/conf.d/default.conf:20]: Yes, I'll leave it in for completeness (to show the page can be served for other error codes) but they probably are redundant and will never be invoked based on the existing known use cases.

File: .github/workflows/pull_request.yml (hunk 4 above)
Comment (pricem14pc) [reply .github/workflows/pull_request.yml]: Removed Dockerhub push

**Review by MebinAbraham — APPROVED** [review]
FYI, on merge worth enabling the push to GAR check in the branch settings for new PRs.

**Pull request merged**